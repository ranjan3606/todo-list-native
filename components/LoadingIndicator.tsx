import React, { useEffect, useRef } from 'react';
import { View, StyleSheet, Animated, Easing } from 'react-native';
import { Colors } from '@/constants/Colors';
import { useActualColorScheme } from '@/services/theme';

interface LoadingIndicatorProps {
  color?: string;
  size?: number;
  spacing?: number;
  fullScreen?: boolean;
  backgroundColor?: string;
}

export const LoadingIndicator: React.FC<LoadingIndicatorProps> = ({ 
  size = 8, 
  spacing = 4,
  color,
  fullScreen = false,
  backgroundColor,
}) => {
  const colorScheme = useActualColorScheme();
  const dotColor = color || Colors[colorScheme].tint;
  const overlayColor = backgroundColor || 'rgba(0,0,0,0.3)';
  
  const dot1 = useRef(new Animated.Value(0)).current;
  const dot2 = useRef(new Animated.Value(0)).current;
  const dot3 = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    const animateDot = (dot: Animated.Value, delay: number) => {
      return Animated.sequence([
        Animated.delay(delay),
        Animated.timing(dot, {
          toValue: 1,
          duration: 400,
          easing: Easing.ease,
          useNativeDriver: true,
        }),
        Animated.timing(dot, {
          toValue: 0,
          duration: 400,
          easing: Easing.ease,
          useNativeDriver: true,
        })
      ]);
    };

    const animateSequence = () => {
      Animated.loop(
        Animated.parallel([
          animateDot(dot1, 0),
          animateDot(dot2, 150),
          animateDot(dot3, 300),
        ])
      ).start();
    };

    animateSequence();
    
    return () => {
      dot1.stopAnimation();
      dot2.stopAnimation();
      dot3.stopAnimation();
    };
  }, [dot1, dot2, dot3]);

  const translateY1 = dot1.interpolate({
    inputRange: [0, 1],
    outputRange: [0, -10]
  });

  const translateY2 = dot2.interpolate({
    inputRange: [0, 1],
    outputRange: [0, -10]
  });

  const translateY3 = dot3.interpolate({
    inputRange: [0, 1],
    outputRange: [0, -10]
  });

  const containerComponent = (
    <View style={styles.container}>
      <Animated.View 
        style={[
          styles.dot, 
          { 
            width: size, 
            height: size, 
            borderRadius: size / 2,
            backgroundColor: dotColor,
            marginHorizontal: spacing,
            transform: [{ translateY: translateY1 }] 
          }
        ]} 
      />
      <Animated.View 
        style={[
          styles.dot, 
          { 
            width: size, 
            height: size, 
            borderRadius: size / 2,
            backgroundColor: dotColor,
            marginHorizontal: spacing,
            transform: [{ translateY: translateY2 }] 
          }
        ]} 
      />
      <Animated.View 
        style={[
          styles.dot, 
          { 
            width: size, 
            height: size, 
            borderRadius: size / 2,
            backgroundColor: dotColor,
            marginHorizontal: spacing,
            transform: [{ translateY: translateY3 }] 
          }
        ]} 
      />
    </View>
  );

  // If fullScreen is true, wrap with overlay
  if (fullScreen) {
    return (
      <View style={[styles.fullScreenOverlay, { backgroundColor: overlayColor }]}>
        {containerComponent}
      </View>
    );
  }

  return containerComponent;
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    padding: 8,
  },
  dot: {
    opacity: 0.8,
  },
  fullScreenOverlay: {
    position: "absolute",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    justifyContent: "center",
    alignItems: "center",
    zIndex: 9999,
  }
});
