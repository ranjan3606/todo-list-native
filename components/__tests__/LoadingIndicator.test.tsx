import React from 'react';
import { render } from '@testing-library/react-native';
import { LoadingIndicator } from '../LoadingIndicator';

jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { tint: '#007bff' },
    dark: { tint: '#00ffcc' },
  }
}));

jest.mock('@/services/theme', () => ({
  useActualColorScheme: () => 'light',
}));

describe('LoadingIndicator', () => {
  it('renders correctly (snapshot)', () => {
    const { toJSON } = render(<LoadingIndicator />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with custom color and size', () => {
    const { toJSON } = render(<LoadingIndicator color="#ff0000" size={16} spacing={8} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders in fullScreen mode', () => {
    const { toJSON } = render(<LoadingIndicator fullScreen backgroundColor="#123456" />);
    expect(toJSON()).toMatchSnapshot();
  });
});
