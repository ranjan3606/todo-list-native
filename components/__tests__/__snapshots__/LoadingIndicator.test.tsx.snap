// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIndicator renders correctly (snapshot) 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "padding": 8,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#007bff",
        "borderRadius": 4,
        "height": 8,
        "marginHorizontal": 4,
        "opacity": 0.8,
        "transform": [
          {
            "translateY": 0,
          },
        ],
        "width": 8,
      }
    }
  />
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#007bff",
        "borderRadius": 4,
        "height": 8,
        "marginHorizontal": 4,
        "opacity": 0.8,
        "transform": [
          {
            "translateY": 0,
          },
        ],
        "width": 8,
      }
    }
  />
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#007bff",
        "borderRadius": 4,
        "height": 8,
        "marginHorizontal": 4,
        "opacity": 0.8,
        "transform": [
          {
            "translateY": 0,
          },
        ],
        "width": 8,
      }
    }
  />
</View>
`;

exports[`LoadingIndicator renders in fullScreen mode 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 9999,
      },
      {
        "backgroundColor": "#123456",
      },
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "padding": 8,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#007bff",
          "borderRadius": 4,
          "height": 8,
          "marginHorizontal": 4,
          "opacity": 0.8,
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 8,
        }
      }
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#007bff",
          "borderRadius": 4,
          "height": 8,
          "marginHorizontal": 4,
          "opacity": 0.8,
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 8,
        }
      }
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#007bff",
          "borderRadius": 4,
          "height": 8,
          "marginHorizontal": 4,
          "opacity": 0.8,
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 8,
        }
      }
    />
  </View>
</View>
`;

exports[`LoadingIndicator renders with custom color and size 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "padding": 8,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#ff0000",
        "borderRadius": 8,
        "height": 16,
        "marginHorizontal": 8,
        "opacity": 0.8,
        "transform": [
          {
            "translateY": 0,
          },
        ],
        "width": 16,
      }
    }
  />
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#ff0000",
        "borderRadius": 8,
        "height": 16,
        "marginHorizontal": 8,
        "opacity": 0.8,
        "transform": [
          {
            "translateY": 0,
          },
        ],
        "width": 16,
      }
    }
  />
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#ff0000",
        "borderRadius": 8,
        "height": 16,
        "marginHorizontal": 8,
        "opacity": 0.8,
        "transform": [
          {
            "translateY": 0,
          },
        ],
        "width": 16,
      }
    }
  />
</View>
`;
